### PR TITLE
ratelimiter for route

### PR DESCRIPTION
**Pull request by qzhuyan**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -140,8 +140,21 @@ It should be set table and the record should be in following spec:
 If file does not exists, brucke will create empty one and use it.
 default: "/tmp/brucke_offsets.DETS"
 
+## ratelimit
+Ratelimit the number of messages produced to downstream.
+
+`ratelimit_interval`  defines the interval,  default: 0 means disabled.
```
Comment (zmstone) [inline README.md]: be clear about milli-seconds?

File: README.md [hunk 2]
```diff
@@ -140,8 +140,21 @@ It should be set table and the record should be in following spec:
...
+`ratelimit_threshold` defines the threshold, set to 0 to pause.
+
+You can change these two settings in realtime via restapi.
+
 example:
+```
+# change ratelimit to 10/s
+POST /route/$routeid/ratelimiter
```
Comment (zmstone) [inline README.md]: explain `routeid`?

File: src/brucke_http_healthcheck_handler.erl [hunk 3]
```diff
@@ -1,5 +1,5 @@
 %%%
-%%%   Copyright (c) 2017 Klarna AB
+%%%   Copyright (c) 2018 Klarna Bank AB (publ)
```
Comment (zmstone) [inline src/brucke_http_healthcheck_handler.erl]: 2017-2018

File: src/brucke_subscriber.erl [hunk 4]
```diff
@@ -52,8 +52,26 @@ start_link(Route, UpstreamPartition, BeginOffset) ->
...
+   , ratelimit_interval := RatelimitInteval
+   , ratelimit_threshold := RatelimitThreshold
    } = Options,
+
   UpstreamClusterName = brucke_config:get_cluster_name(UpstreamClientId),
+
+  RatelimiterName = brucke_ratelimiter:name_it([binary_to_list(UpstreamClusterName), binary_to_list(CgId)]),
+  Ratelimiter = case RatelimitInteval =/= ?RATELIMIT_DISABLED of
+                  true ->
+                    case brucke_ratelimiter:start_link(RatelimiterName, {RatelimitInteval, RatelimitThreshold}) of
```
Comment (zmstone) [inline src/brucke_subscriber.erl]: maybe add a `brucke_ratelimter:ensure_started`

**Comment by qzhuyan** [comment]
There are two features would be nice to have:
1) add ability to PUT/DELETE `ratelimiter` via rest api for existing `route`
2) add http auth for security

File: src/brucke_ratelimiter.erl [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+  {reply, {Interval, Threshold}, State};
+
+handle_call({set_rate, Rateprop}, _From, #state{ ticker = OldTref
+                                               , id = Rid
+                                               , interval = OldInterval
+                                               , threshold = OldThreshold} = State) ->
+  NewInterval  = proplists:get_value(interval, Rateprop, undefined),
+  NewThreshold = proplists:get_value(threshold, Rateprop, undefined),
+
+  %%% fine contorl what to reset to reduce unnecessary jitter
```
Comment (k32) [inline src/brucke_ratelimiter.erl]: s/contorl/control/

File: src/brucke_ratelimiter_http_handler.erl [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+handle_request(Req, State) ->
+  {Cluster, Req1} = cowboy_req:binding(cluster, Req),
+  {Cgid, Req1} = cowboy_req:binding(cgid, Req),
+  {ok, Body, Req2} = cowboy_req:body(Req1),
+  Args = lists:map(fun({<< "interval" >>, V}) ->
+                       {interval, to_int(V)};
+                      ({<< "threshold" >>, V}) ->
+                       {threshold, to_int(V)}
+                   end, jsone:decode(Body, [{object_format,proplist}])),
+  Res = brucke_ratelimiter:set_rate({Cluster, Cgid}, Args),
```
Comment (k32) [inline src/brucke_ratelimiter_http_handler.erl:48]: We probably want some sort of authentication.

Comment (qzhuyan) [reply src/brucke_ratelimiter_http_handler.erl:48]: sure, that might be in next PR.

**Pull request merged**